google_compute_engine_oslogin/pam_module/pam_oslogin_login.cc

```diff
@@ -108,4 +114,110 @@ PAM_EXTERN int pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc,
   }
   return pam_result;
 }
+PAM_EXTERN int pam_sm_setcred(pam_handle_t * pamh, int flags, int argc,
+                                              const char **argv)
+{
+            return PAM_SUCCESS;
+}
+
+
+PAM_EXTERN int pam_sm_authenticate(pam_handle_t * pamh, int flags,
+                                                   int argc, const char **argv)
+{
+  const char* user_name;
+  if (pam_get_user(pamh, &user_name, NULL) != PAM_SUCCESS) {
+    pam_syslog(pamh, LOG_INFO, "Could not get pam user.");
+    return PAM_PERM_DENIED;
+  }
+
+  string str_user_name(user_name);
+  if (!ValidateUserName(user_name))
+    return PAM_PERM_DENIED;
+
+  string response;
+  if (!(GetUser(str_user_name, &response)))
+    return PAM_PERM_DENIED;
+
+  string email;
+  if (!ParseJsonToEmail(response, &email) || email.empty())
+    return PAM_PERM_DENIED;
+
+  response = "";
+  if (!StartSession(email, &response)) {
+    pam_syslog(pamh, LOG_ERR, "Bad response from 2-factor start API: %s",
+               response.empty() ? "empty response" : response.c_str());
+    return PAM_PERM_DENIED;
+  }
+
+  string status;
+  if (!ParseJsonToKey(response, "status", &status)) {
+    pam_syslog(pamh, LOG_ERR, "Failed to parse status from start session response");
+    return PAM_PERM_DENIED;
+  }
+
+  if (status == "NO_AVAILABLE_CHALLENGES")
+    return PAM_SUCCESS; // user is not 2-factor enabled
+
+  string session_id;
+  if (!ParseJsonToKey(response, "sessionId", &session_id))
```

Comment:
Optional, but I prefer (as a general rule) to always have a curly brace with an if statement. It prevents pain later on.

Reply:
+1, also maintains style consistency.